src/tiled.c

```diff
@@ -1,19 +1,55 @@
 #include "../include/tiled.h"
+#include "../include/format_handling.h"
+#include "../include/csv_reader.h"
+#include "../include/memory_alloc.h"
 
 #include <string.h>
 
 #define FILE_BUFFER_SIZE 100
 
+#define CHECK_BOUNDS()                                                                             \
+	if (!is_intersecting(screen_rect, tiled->map_rect))                                            \
+		return;
+
 #define SET_VARS()                                                                                 \
-	size_t initial_x = screen_rect.pos.x / tiled->tile_size.width;                                 \
-	size_t initial_y = screen_rect.pos.y / tiled->tile_size.height;                                \
-	size_t final_x = ((screen_rect.pos.x + screen_rect.size.width) / tiled->tile_size.width) + 1;  \
-	size_t final_y = ((screen_rect.pos.y + screen_rect.size.height) / tiled->tile_size.height) + 1;
+	int initial_x = (screen_rect.pos.x - tiled->offset.x - view_position.x) /                      \
+					tiled->tile_size.width;                                                        \
+	int initial_y = (screen_rect.pos.y - tiled->offset.y - view_position.y) /                      \
+					tiled->tile_size.height;                                                       \
+	size_t final_x = ((screen_rect.pos.x + screen_rect.size.width - tiled->offset.x -              \
```

Comment:
Since the `initial_x` and `initial_y` values can now be negative, should we do the same for `final_x` and `final_y`?